File: 백준/Silver/2644. 촌수계산/촌수계산.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
void bfs(int **arr, int *arr_cnt, int *visited, int start, int end, int n) {
    visited[start] = 0;
    int idx = 0, node, *queue = (int *)malloc(sizeof(int) * n);
    queue[idx++] = start;
    while (idx > 0) {
        node = queue[0];
        for (int i = 1; i < idx; i++) {
            queue[i - 1] = queue[i];
        }
        idx--;
        if (node == end) return;
        for (int i = 0; i < arr_cnt[node]; i++) {
            if (visited[arr[node][i]] == -1) {
                visited[arr[node][i]] = visited[node] + 1;
                queue[idx++] = arr[node][i];
            }
        }
    }
}
int main(void) {
    int n, m, start, end, u, v;
    scanf("%d", &n);
    scanf("%d %d", &start, &end);
    int *visited = (int *)malloc(sizeof(int) * (n + 1));
    memset(visited, -1, sizeof(int) * (n + 1));
    int **arr = (int **)malloc(sizeof(int *) * (n + 1));
    for (int i = 0; i < n + 1; i++)
        arr[i] = (int *)malloc(sizeof(int) * n);
    int *arr_cnt = (int *)malloc(sizeof(int) * (n + 1));
    memset(arr_cnt, 0, sizeof(int) * (n + 1));
    scanf("%d", &m);
    for (int i = 0; i < m; i++) {
        scanf("%d %d", &u, &v);
        arr[u][arr_cnt[u]++] = v;
        arr[v][arr_cnt[v]++] = u;
    }
    bfs(arr, arr_cnt, visited, start, end, n);
    printf("%d", visited[end]);
    return 0;
}

File: 백준/Silver/2644. 촌수계산/촌수계산.py
import sys
from collections import deque
input=lambda:sys.stdin.readline().rstrip()
n=int(input())
visited=[-1]*(n+1)
arr=[[] for _ in range(n+1)]
start,end=map(int,input().split())
m=int(input())
for i in range(m):
    u,v=map(int,input().split())
    arr[u]+=[v]
    arr[v]+=[u]
def bfs(st):
    q=deque([st])
    visited[st]=0
    r=0
    while q:
        node=q.popleft()
        if node==end:
            return
        r+=1
        for nn in arr[node]:
            if visited[nn]==-1:
                visited[nn]=visited[node]+1;
                q+=[nn]
bfs(start)
print(visited[end])